src/types/array.js

```diff
@@ -48,9 +48,11 @@ export default parameterized(T => class ArrayType {
   }
 
   map(fn) {
-    return valueOf(this).map((item, index) => {
-      return valueOf(fn(create(T, item)));
-    });
+    let list = valueOf(this);
+    return list.reduce((acc, item, index) => {
```

Comment:
I actually prefer not to use a generic variable like `acc` but instead use something with meaning in the context, like `values`.  Yes it is an accumulator in the context of an abstract reduction, but this is not an abstract reduction, it's a concrete reduction and the variable references very specific content: _the underlying array of values of the resulting microstate_.